### PR TITLE
reword #[test] attribute error on fn items

### PR DESCRIPTION
**Pull request by zeegomo**
fix of [#55787](https://github.com/rust-lang/rust/issues/55787)
Reworded error message from "#[test] attribute is only allowed on fn items" to "#[test] attribute is only allowed on non associated functions"

### CODE REVIEW

**Comment by estebank** [comment]
Can you add a test? Other than that, r=me.

**Comment by Zeegomo** [comment]
sure, where should I add it? I can't find other tests of that file

**Comment by estebank** [comment]
Create a file in `src/test/ui/` with a repro case and run `./x.py test src/test/ui --stage 1 --bless` to create the `.stderr`. Add both files to git and push the change.

File: src/libsyntax_ext/test.rs [hunk 1]
```diff
@@ -53,7 +53,7 @@ pub fn expand_test_or_bench(
         if let Annotatable::Item(i) = item { i }
         else {
             cx.parse_sess.span_diagnostic.span_fatal(item.span(),
-                "#[test] attribute is only allowed on fn items").raise();
+                "#[test] attribute is only allowed on non associated functions").raise();
```
Comment (Centril) [inline src/libsyntax_ext/test.rs]: I would rephrase `non associated functions` as `free functions` to avoid the negative.

Comment (Zeegomo) [reply src/libsyntax_ext/test.rs]: a simple search of "rust free functions" on duckduckgo or "free functions" on the rust book has no results while "associated functions" gives you the expected page, that's why I chose that phrase.

Comment (Centril) [reply src/libsyntax_ext/test.rs]: Sadly the [page in the book](https://doc.rust-lang.org/book/2018-edition/ch05-03-method-syntax.html) is [quite wrong](https://doc.rust-lang.org/stable/reference/items/associated-items.html) in that it defines *"associated function"* as a non-method inside an implementation (trait/inherent). Thus, by saying "non-associated function", by the incorrect definition in the book you are stating that `#[test]` is allowed on methods. Given this, I think if you want to impart a correct error you need to say something like:

> `"#[test] attribute is only allowed on free functions without arguments"`

or:

> `"#[test] attribute is only allowed on no-argument functions outside implementations"`
